flink-streaming-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer010.java

```diff
@@ -0,0 +1,198 @@
+
+	/**
+	 * Creates a FlinkKafkaProducer for a given topic. The sink produces a DataStream to
+	 * the topic.
+	 *
+	 * @param inStream The stream to write to Kafka
+	 * @param topicId ID of the Kafka topic.
+	 * @param serializationSchema User defined serialization schema supporting key/value messages
+	 * @param producerConfig Properties with the producer configuration.
+	 */
+	public static <T> FlinkKafkaProducer010Configuration writeToKafka(DataStream<T> inStream,
+										String topicId,
+										KeyedSerializationSchema<T> serializationSchema,
+										Properties producerConfig) {
+		return writeToKafka(inStream, topicId, serializationSchema, producerConfig, new FixedPartitioner<T>());
+	}
+
+
+	/**
+	 * Creates a FlinkKafkaProducer for a given topic. the sink produces a DataStream to
+	 * the topic.
+	 *
+	 * @param inStream The stream to write to Kafka
+	 * @param topicId ID of the Kafka topic.
+	 * @param serializationSchema User defined (keyless) serialization schema.
+	 * @param producerConfig Properties with the producer configuration.
+	 */
+	public static <T> FlinkKafkaProducer010Configuration writeToKafka(DataStream<T> inStream,
+										String topicId,
+										SerializationSchema<T> serializationSchema,
+										Properties producerConfig) {
+		return writeToKafka(inStream, topicId, new KeyedSerializationSchemaWrapper<>(serializationSchema), producerConfig, new FixedPartitioner<T>());
+	}
+
+	/**
+	 * Creates a FlinkKafkaProducer for a given topic. The sink produces a DataStream to
+	 * the topic.
+	 *  @param inStream The stream to write to Kafka
+	 *  @param topicId The name of the target topic
+	 *  @param serializationSchema A serializable serialization schema for turning user objects into a kafka-consumable byte[] supporting key/value messages
+	 *  @param producerConfig Configuration properties for the KafkaProducer. 'bootstrap.servers.' is the only required argument.
+	 *  @param customPartitioner A serializable partitioner for assigning messages to Kafka partitions.
+	 */
+	public static <T> FlinkKafkaProducer010Configuration<T> writeToKafka(DataStream<T> inStream,
+																String topicId,
+																KeyedSerializationSchema<T> serializationSchema,
+																Properties producerConfig,
+																KafkaPartitioner<T> customPartitioner) {
+		GenericTypeInfo<Object> objectTypeInfo = new GenericTypeInfo<>(Object.class);
+		FlinkKafkaProducer010<T> kafkaProducer = new FlinkKafkaProducer010<>(topicId, serializationSchema, producerConfig, customPartitioner);
+		SingleOutputStreamOperator<Object> transformation = inStream.transform("FlinKafkaProducer 0.10.x", objectTypeInfo, kafkaProducer);
+		return new FlinkKafkaProducer010Configuration<>(transformation, kafkaProducer);
+	}
+
+	/**
+	 * Configuration object returned by the writeToKafka() call.
+	 */
+	public static class FlinkKafkaProducer010Configuration<T> extends DataStreamSink<T> {
```

Comment:
I wonder if we can move the config setter methods into `FlinkKafkaProducer010`, and let `writeToKafka` return the created 0.10 producer instead of instantiating a new `FlinkKafkaProducer010Configuration`? My reasoning is that I find it a bit strange to set these config values on a separate config object, but not "applying" it anywhere. I think users would normally expect they have to "apply" a completed config object somewhere to take effect. So, I have something like this in mind:

In constructor of `FlinkKafkaProducer010`
```
private FlinkKafaProducer010(...) {
  super(new FlinkKafkaProducer09<>(...));
  this.wrapped09Producer = (FlinkKafkaProducer09<T>) userFunction; // wrapped09Producer as FlinkKafkafProducer010's class field
}
```
Then FlinkKafkaProducer010 can directly have the setter methods:
```
public void setFlushOnCheckpoint(boolean flush) {
    this.wrapped09producer.setFlushOnCheckpoint(flush);
}

public void setLogFailuresOnly(boolean logFailuresOnly) {
	this.wrapped09producer.setLogFailuresOnly(logFailuresOnly);
}

public void setWriteTimestampToKafka(boolean writeTimestampToKafka) {
	this.writeTimestampToKafka = writeTimestampToKafka;
}
```

Then users can use the producer like this:
```
FlinkKafkaProducer010 producer = FlinkKafkaProducer010.writeToKafka(...)
producer.setLogFailuresOnly(...)
producer.setFlushOnCheckpoint(...)
producer.setWriteTimestampToKafka(...)
```

Although still a bit different from the usage patterns of previous versions, it looks a bit more similar compared to what the PR has now. I'm not sure if I may be missing something that lead to choosing to have a separate `FlinkKafkaProducer010Configuration`. Is there any particular reason?